### PR TITLE
Refactor PauliNet core

### PR DESCRIPTION
**Pull request by zenoone**
This pull request refactores the core of `PauliNet`, moving the evaluation of the Jastrow factor and the backflow transformation to `OmniSchNet` and establishes mean-field variants of the Jastrow factor and backflow transformation, by introducing a mean-field version of the `ElectronicSchNet`.

### CODE REVIEW

File: src/deepqmc/wf/paulinet/backflow.py [hunk 1]
```diff
@@ -1,38 +1,26 @@
...
-    def __init__(self, mol, embedding_dim, subnets_factory=None):
-        if not subnets_factory:
-
-            def subnets_factory(embedding_dim):
-                return (
-                    get_log_dnn(embedding_dim, 1, SSP, n_layers=3),
-                    get_log_dnn(embedding_dim, len(mol), SSP, n_layers=3),
-                )
-
+    def __init__(
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/backflow.py:8]: I would actually move this to `omni.py`, because it's a really small class and is only used there. Also, this backflow is entirely different from the one that was here before.

File: src/deepqmc/wf/paulinet/jastrow.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from torch import nn
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/jastrow.py:1]: Same as for `Backflow` above.

File: src/deepqmc/wf/paulinet/paulinet.py [hunk 3]
```diff
@@ -173,25 +153,18 @@ def __init__(
             backflow_spec[1] *= 2
         self.backflow_type = backflow_type
         self.backflow_transform = backflow_transform
-        self.backflow = (
-            backflow_factory(len(mol), dist_feat_dim, n_up, n_down, *backflow_spec)
-            if backflow_factory
+        self.omni = (
+            omni_factory(mol, n_up, n_down, backflow_spec, **omni_kwargs)
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/paulinet.py:157]: Does omni actually need the full molecule object? Isn't its size enough?

File: src/deepqmc/wf/paulinet/paulinet.py [hunk 4]
```diff
@@ -173,25 +153,18 @@ def __init__(
...
-            assert not backflow_factory and not jastrow_factory
-            self.omni = omni_factory(
-                mol, dist_feat_dim, n_up, n_down, *backflow_spec, **(omni_kwargs or {})
-            )
-            self.backflow = self.omni.forward_backflow
-            self.r_backflow = self.omni.forward_r_backflow
-            self.jastrow = self.omni.forward_jastrow
-        else:
-            self.omni = None
+        self.jastrow = (
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/paulinet.py:161]: I'm thinking whether we shouldn't simplify this, now that `jastrow` and `backflow` will always refer to the functions from `omni`. Not sure

File: src/deepqmc/wf/paulinet/paulinet.py [hunk 5]
```diff
@@ -301,7 +272,14 @@ def from_pyscf(
             mf.mol.spin,
         )
         basis = GTOBasis.from_pyscf(mf.mol)
-        wf = cls(mol, basis, **{'omni_factory': OmniSchNet, **kwargs})
+        wf = cls(
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/paulinet.py:275]: This change isn't needed, you need to remove the trailing comma, then black will collapse it.

File: src/deepqmc/wf/paulinet/paulinet.py [hunk 6]
```diff
@@ -438,8 +407,6 @@ def forward(self, rs):  # noqa: C901
                 else psi * torch.exp(cusp_same + cusp_anti)
             )
         if self.jastrow:
-            J = self.jastrow(*edges)
+            J = self.jastrow(dists_nuc, dists_elec)
             psi = psi + J if self.return_log else psi * torch.exp(J)
-        if self.omni:
-            self.omni.forward_close()
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/paulinet.py:444]: Are we certain this isn't needed? I thought it shouldn't be needed, but then randomly, I’d get crashes, apparently the ID was actually the same for a different tensor. Now that I think of it, perhaps because we use the sum of the two id for the total id? Maybe a tuple of them would actually work.

File: src/deepqmc/wf/paulinet/r_backflow.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import torch
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/r_backflow.py:1]: I'd move this to `deepqmc.extra` actually, since we don't really test this code.

File: src/deepqmc/wf/paulinet/schnet.py [hunk 8]
```diff
@@ -200,26 +211,28 @@ class ElectronicSchNet(nn.Module):
...
         kernel_dim=64,
+        embedding_dim=64,
         version=2,
         layer_norm=False,
     ):
         assert version in self.layer_factories
         subnet_metafactory = subnet_metafactory or SubnetFactory
         subnet_factory = subnet_metafactory(dist_feat_dim, kernel_dim, embedding_dim)
         super().__init__()
+        self.embedding_dim = embedding_dim
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/schnet.py:235]: This shouldn't be here, since it's not used by the actual object. If some upstream object needs it, the embedding dim should be passed down from that upstream object

File: tests/test_wfnet.py [hunk 9]
```diff
@@ -69,7 +69,6 @@ def wf(net_factory, mol):
         mole = pyscf.gto.M(atom=mol.as_pyscf(), unit='bohr', basis='6-311g', cart=True)
         basis = GTOBasis.from_pyscf(mole)
         args += (basis,)
-        kwargs.update({'jastrow_factory': JastrowNet, 'dist_feat_dim': 4})
```
Comment (jhrmnn) [inline tests/test_wfnet.py:72]: I think with this removal, the test tests only a non-trainable ansatz, isn't that right? This should be rather replaced by some stripped-down version of the omni.

File: src/deepqmc/wf/paulinet/omni.py [hunk 10]
```diff
@@ -1,158 +1,161 @@
...
+class OmniCore(nn.Module):
     def __init__(
         self,
         mol,
-        dist_feat_dim,
         n_up,
         n_down,
-        n_orbitals,
-        n_channels,
+        backflow_spec,
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/omni.py:18]: It would be more readable if this was passed through unrolled.

File: src/deepqmc/wf/paulinet/omni.py [hunk 11]
```diff
@@ -1,158 +1,161 @@
...
-        n_jastrow_layers=3,
-        with_backflow=True,
-        n_backflow_layers=3,
-        with_r_backflow=False,
-        schnet_kwargs=None,
-        subnet_kwargs=None,
+        dist_feat_dim=32,
+        dist_feat_cutoff=10.0,
+        many_body_dnn_factory=ElectronicSchNet,
+        mean_field_dnn_factory=partial(ElectronicSchNet, version='mean-field'),
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/omni.py:23]: I would subclass Schnet for a default mean-field version.

File: src/deepqmc/wf/paulinet/omni.py [hunk 12]
```diff
@@ -1,158 +1,161 @@
...
-            self.forward_jastrow = None
-        if with_backflow:
-            backflow = [
-                get_log_dnn(
-                    embedding_dim,
-                    n_orbitals,
-                    SSP,
-                    n_layers=n_backflow_layers,
-                    last_bias=True,
+        self.dnn_core = nn.ModuleDict(
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/omni.py:35]: The dict is never used as a whole, it's always just its components, so having two separate attributes would be more straightforward.

**Review by jhrmnn — CHANGES_REQUESTED** [review]

File: src/deepqmc/wf/paulinet/omni.py [hunk 13]
```diff
@@ -1,158 +1,161 @@
...
+                else None,
+            }
+        )
+        self.dist_basis = (
+            DistanceBasis(dist_feat_dim, cutoff=dist_feat_cutoff, envelope='nocusp')
+            if 'edges' in jastrow_stream + backflow_stream
+            or 'emb' in jastrow_stream + backflow_stream
+            else None
+        )
+        self.streams = {
```
Comment (jhrmnn) [inline src/deepqmc/wf/paulinet/omni.py:63]: the `streams` dict is only used in the initializer, so it's not needed as an attribute?

**Review by jhrmnn — CHANGES_REQUESTED** [review]

File: src/deepqmc/wf/paulinet/paulinet.py (hunk 4 above)
Comment (jhrmnn) [reply src/deepqmc/wf/paulinet/paulinet.py:161]: I think everything would simplify a lot, if we had a single `forward()` on omni, that would return both the jastrow and the backflow. Then this whole business with the "memory" can be removed, and the stream logic will be much simpler. What do you think?

File: src/deepqmc/wf/paulinet/schnet.py (hunk 8 above)
Comment (zenoone) [reply src/deepqmc/wf/paulinet/schnet.py:235]: That is true, however when a different network is used to generate the embeddings it might not have the same submodules, but it certainly has an embedding dimensions. That is to say this should make the architecture more modular.

File: src/deepqmc/wf/paulinet/paulinet.py (hunk 6 above)
Comment (zenoone) [reply src/deepqmc/wf/paulinet/paulinet.py:444]: Good catch with the sum, this could clearly cause ambiguent IDs. If this works properly I think the forward close shouldn't be needed.

File: src/deepqmc/wf/paulinet/paulinet.py (hunk 4 above)
Comment (zenoone) [reply src/deepqmc/wf/paulinet/paulinet.py:161]: You are right, this is not needed and both could be combined in a single forward call. I thought that this way its more tracable what happens and the forward of the PauliNet does not have to be adapted, though I'd be open for changing it. However I do not see how that would allow the memory to be removed and how it would change the stream logic. Consider the following example: Both Jastrow and backflow are build upon the many-body embeddings. Then I would still have to store the many-body embeddings after the first forward call of the ElectronicSchNet for lets say the Jastrow, to be reused in the backflow. Or another example: Jastrow uses many-body embeddings and backflow uses edges. Then I would like to store the edges in the process of obtaining the many-body representation to be reused when called in the backflow. Did I miss something?

Comment (zenoone) [reply src/deepqmc/wf/paulinet/paulinet.py:161]: Speaking of _the first forward call of the ElectronicSchNet for lets say the **Jastrow**, to be reused in the **backflow**_ I mean Jastrow and backflow calles in omni.py.
